rclcpp/include/rclcpp/topic_statistics/subscriber_topic_statistics.hpp

```diff
@@ -0,0 +1,194 @@
+  : node_name_(node_name),
+    publisher_(std::move(publisher))
+  {
+    // TODO(dbbonnie): ros-tooling/aws-roadmap/issues/226, received message age
+
+    auto rmp = std::make_unique<received_message_period>();
+    rmp->Start();
+    subscriber_statistics_collectors_.emplace_back(std::move(rmp));
+
+    window_start_ = rclcpp::Time(GetCurrentNanosecondsSinceEpoch());
+  }
+
+  virtual ~SubscriberTopicStatistics()
+  {
+    TearDown();
+  }
+
+  /// Stop all collectors, clear measurements, stop publishing timer, and reset publisher.
+  virtual void TearDown()
+  {
+    for (auto & collector : subscriber_statistics_collectors_) {
+      collector->Stop();
+    }
+
+    subscriber_statistics_collectors_.clear();
+
+    if (publisher_timer_) {
+      publisher_timer_->cancel();
+      publisher_timer_.reset();
+    }
+    if (publisher_) {
+      publisher_.reset();
+    }
+  }
+
+  /// Set the timer used to publish statistics messages.
+  /**
+   * @param measurement_timer the timer to fire the publisher, created by the node
+   */
+  void SetPublisherTimer(const rclcpp::TimerBase::SharedPtr & publisher_timer)
+  {
+    publisher_timer_ = std::move(publisher_timer);
+  }
+
+  /// Handle a message received by the subscription to collect statistics.
+  /**
+   * @param received_message the message received by the subscription
+   * @param now_nanoseconds current time in nanoseconds
+   */
+  virtual void OnMessageReceived(
+    const CallbackMessageT & received_message,
+    const rcl_time_point_value_t now_nanoseconds) const
+  {
+    (void) received_message;
+
+    for (auto & collector : subscriber_statistics_collectors_) {
+      collector->OnMessageReceived(received_message, now_nanoseconds);
+    }
+  }
+
+  /// Return a vector of all the currently collected data
+  /**
+   * @return a vector of all the collected data
+   */
+  std::vector<StatisticData> GetCurrentCollectorData() const
+  {
+    std::vector<StatisticData> data;
+    for (const auto & collector : subscriber_statistics_collectors_) {
+      data.push_back(collector->GetStatisticsResults());
+    }
+    return data;
+  }
+
+private:
+  /// Publish a populated MetricsStatisticsMessage
+  virtual void PublishMessage()
+  {
+    rclcpp::Time window_end{GetCurrentNanosecondsSinceEpoch()};
+
+    for (auto & collector : subscriber_statistics_collectors_) {
+      const auto collected_stats = collector->GetStatisticsResults();
+
+      auto message = libstatistics_collector::collector::GenerateStatisticMessage(
+        node_name_,
+        collector->GetMetricName(),
+        collector->GetMetricUnit(),
+        window_start_,
+        window_end,
+        collected_stats);
+      publisher_->publish(message);
+    }
+    window_start_ = window_end;
+  }
+
+  /// Return the current nanoseconds (count) since epoch
+  /**
+   * Based on design discussions, using harccoded time instead of a node's clock
+   * due to lifecycle issues.
```

Comment:
Which design discussions?

Reply:
Sorry, this was an email thread with @wjwwood .